### PR TITLE
Implements the provider side of sdcore management interface

### PR DESCRIPTION
**Pull request by saltiyazan**
# Description

Implements the provider side of sdcore management interface and adds unit tests.

# Checklist:

- [x] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that validate the behaviour of the software
- [x] I validated that new and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [ ] I have bumped the version of the library

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -26,6 +32,22 @@
...
+
+def _get_pod_ip() -> Optional[str]:
+    """Returns the pod IP using juju client.
+
+    Returns:
+        str: The pod IP.
+    """
+    try:
+        ip_address = check_output(["unit-get", "private-address"])
+        print(ip_address)
```
Comment (ghislainbourgeois) [inline src/charm.py:47]:
```suggestion
```

File: src/charm.py [hunk 2]
```diff
@@ -128,10 +183,20 @@ def _relation_created(self, relation_name: str) -> bool:
             relation_name (str): Relation name
 
         Returns:
-            str: Whether the relation was created.
+            bool: Whether the relation was created.
         """
         return bool(self.model.get_relation(relation_name))
 
+    def _get_webui_endpoint_url(self) -> Optional[str]:
```
Comment (ghislainbourgeois) [inline src/charm.py:190]:
```suggestion
    def _get_webui_endpoint_url(self) -> str:
```
This always return a string.

**Review by ghislainbourgeois — CHANGES_REQUESTED** [review]
Some small nitpicks

File: src/charm.py (hunk 1 above)
Comment (saltiyazan) [reply src/charm.py]: Oh, removed

File: src/charm.py (hunk 2 above)
Comment (saltiyazan) [reply src/charm.py:189]: ✅  returning None now

File: src/charm.py [hunk 3]
```diff
@@ -105,6 +143,23 @@ def _on_database_created(self, event: DatabaseCreatedEvent) -> None:
         self._write_config_file(content=config_file_content)
         self._on_webui_pebble_ready(event=event)
 
+    def _publish_sdcore_management_url(self, event: EventBase):
+        """Handles sdcore management relation joined event.
```
[the remark this answers is not in the document]
Comment (saltiyazan) [reply src/charm.py]: Correct, fixed

File: src/charm.py [hunk 4]
```diff
@@ -128,10 +183,20 @@ def _relation_created(self, relation_name: str) -> bool:
             relation_name (str): Relation name
 
         Returns:
-            str: Whether the relation was created.
+            bool: Whether the relation was created.
```
[the remark this answers is not in the document]
Comment (saltiyazan) [reply src/charm.py:185]: That's true, maybe we should refactor this, not only here but in the rest of the projects too, I'd probably bring that into a separate discussion, what do you think?

[the remark this answers is not in the document]
Comment (DanielArndt) [reply src/charm.py:185]: Yep that sounds like a more appropriate way to deal with it.

**Review by DanielArndt — APPROVED** [review]

**Review by ghislainbourgeois — APPROVED** [review]

**Pull request merged**